exec/supervisor/plugin.go

```diff
@@ -179,24 +194,33 @@ func (p *Plugin) execute(program *Program) error {
 	p.programs[program.Name] = &processWithStateChan{
 		process:   process,
 		stateChan: stateChan,
+		doneChan:  doneChan,
 		svLogger:  svLogger,
 	}
 
 	return nil
 }
 
-func (p *Plugin) watch(stateChan chan status.ProcessStatus, name string) {
+func (p *Plugin) watch(stateChan chan status.ProcessStatus, doneChan chan struct{}, name string) {
+	p.wg.Add(1)
```

Comment:
I've mentioned this few times before, but to be more specific I will quote important part of documentation for [`WaitGroup.Add`](https://godoc.org/sync#WaitGroup.Add): 

> Typically this means the calls to Add should execute before the statement creating the goroutine or other event to be waited for...

Reply:
Agree, but the reason for this rule is to prevent calling `Wait()` before the first `Add()` which may happen when the goroutine execution is delayed.

Reply:
That is just wrong. You should realize the flow of the plugin. You have chain of goroutines starting from `Init` method and continuing `startPrograms` and then this `watch` where you can never ensure the the time of execution. The `Close` could be called immediately after `Init` when the `p.programs` is still empty, which can easily happen, because of so many unsynchronized goroutines, which would actually make the execution skip to the `Wait` call and immediately exiting, possibly creating a lot of race conditions here. 

This is simply bad practice that breeds unreliable code.